crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,352 @@
+        Hint ratings in hint_database are updated and the resulting hint rating (or reported status) is returned to JS.
+        Args:
+          data['student_answer']: The incorrect answer that corresponds to the hint that is being rated
+          data['hint']: The hint that is being rated
+          data['student_rating']: The rating chosen by the student.
+        Returns:
+          'rating': the new rating of the hint, or the string 'reported' if the hint was reported
+          'hint': the hint that had its rating changed
+        """
+        answer_data = data['student_answer']
+        data_rating = data['student_rating']
+        data_hint = data['hint']
+        if data_hint == 'Sorry, there are no hints for this answer.':
+            return {"rating": None, 'hint': data_hint}
+        if data['student_rating'] == 'unreport':
+            for reported_hints in self.reported_hints:
+                if reported_hints == data_hint:
+                    self.reported_hints.pop(data_hint, None)
+                    return {'rating': 'unreported'}
+        if data['student_rating'] == 'remove':
+            for reported_hints in self.reported_hints:
+                if data_hint == reported_hints:
+                    self.hint_database[self.reported_hints[data_hint]].pop(data_hint, None)
+                    self.reported_hints.pop(data_hint, None)
+                    return {'rating': 'removed'}
+        if data['student_rating'] == 'report':
+            # add hint to Reported dictionary
+            self.reported_hints[str(data_hint)] = answer_data
+            return {"rating": 'reported', 'hint': data_hint}
+        rating = self.change_rating(data_hint, data_rating, answer_data)
+        return {"rating": str(rating), 'hint': data_hint}
+
+    def change_rating(self, data_hint, data_rating, answer_data):
+        """
+        This function is used to change the rating of a hint when students vote on its helpfulness.
+        Initiated by rate_hint. The temporary_dictionary is manipulated to be used
+        in self.rate_hint
+        Args:
+          data_hint: This is equal to the data['hint'] in self.rate_hint
+          data_rating: This is equal to the data['student_rating'] in self.rate_hint
+          answer_data: This is equal to the data['student_answer'] in self.rate_hint
+        Returns:
+          The rating associated with the hint is returned. This rating is identical
+          to what would be found under self.hint_database[answer_string[hint_string]]
+        """
+        if any(data_hint in generic_hints for generic_hints in self.generic_hints):
+            return
+        if data_rating == 'upvote':
+            self.hint_database[str(answer_data)][str(data_hint)] += 1
+            return self.hint_database[str(answer_data)][str(data_hint)]
+        else:
+            self.hint_database[str(answer_data)][str(data_hint)] -= 1
+            return self.hint_database[str(answer_data)][str(data_hint)]
```

Comment:
Minor: lines 291-295 are really similar.  The returns are the same, and can be moved out of the if/else.  The incrementing and decrementing lines are almost the same.  Use the if to set a delta variable, then use one line to add delta to the hint_database value.